### PR TITLE
[FRONT-1536] Staking

### PR DESCRIPTION
**Pull request by pmichalowski**

### CODE REVIEW

File: src/components/ActionBars/SponsorshipActionBar.tsx [hunk 1]
```diff
@@ -30,14 +32,18 @@ import {
...
 
     const fundSponsorship = useFundSponsorship()
+    const { canJoinSponsorship, joinSponsorship } = useJoinSponsorship()
+    const tokenInfo = useTokenInfo(
+        defaultChainConfig.contracts[getCoreConfig().sponsorshipPaymentToken],
+        defaultChainConfig.id,
+    )
+    const tokenSymbol = tokenInfo?.symbol || 'DATA'
 
     // TODO when Mariusz will merge his hook & getter for fetching Token information - use it here to display the proper token symbol
```
Comment (tumppi) [inline src/components/ActionBars/SponsorshipActionBar.tsx:48]: Comment can be removed?

File: src/hooks/useJoinSponsorship.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        joinSponsorship: async (sponsorshipId, sponsorshipStreamId) => {
+            if (!myOperatorQuery.data) {
+                return
+            }
+            const balance = myOperatorQuery.data?.poolValue.minus(
+                myOperatorQuery.data?.stakes.reduce((prev, curr, index) => {
+                    return prev.plus(curr.amount)
+                }, toBN(0)),
+            )
+            try {
```
Comment (tumppi) [inline src/hooks/useJoinSponsorship.ts:32]: Remove try/catch to make possible errors visible

File: src/hooks/useJoinSponsorship.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    const myOperatorQuery = useMyOperator(wallet || '')
+    const canJoinSponsorship = useMemo(() => !!myOperatorQuery.data, [myOperatorQuery])
+
+    return {
+        canJoinSponsorship,
+        joinSponsorship: async (sponsorshipId, sponsorshipStreamId) => {
+            if (!myOperatorQuery.data) {
+                return
+            }
+            const balance = myOperatorQuery.data?.poolValue.minus(
```
Comment (tumppi) [inline src/hooks/useJoinSponsorship.ts:27]: We should use `operator.freeFundsWei` to get available balance on operator contract.

File: src/modals/JoinSponsorshipModal.tsx [hunk 4]
```diff
@@ -49,17 +58,24 @@ export default function JoinSponsorshipModal({
 
     const [rawAmount, setRawAmount] = useState(parseAmount(amountProp))
 
-    const amount = toBN(rawAmount || '0').multipliedBy(1e18)
+    const amount = toBN(rawAmount || '0').multipliedBy(Math.pow(10, decimals))
```
Comment (tumppi) [inline src/modals/JoinSponsorshipModal.tsx:61]: There is `toDecimals` helper available in utils/math.

File: src/modals/JoinSponsorshipModal.tsx [hunk 5]
```diff
@@ -148,7 +160,8 @@ export default function JoinSponsorshipModal({
                             )}
                         </Prop>
                         <div>
-                            {operatorBalance.dividedBy(1e18).toString()} {tokenSymbol}
+                            {operatorBalance.dividedBy(Math.pow(10, decimals)).toString()}{' '}
```
Comment (tumppi) [inline src/modals/JoinSponsorshipModal.tsx:163]: `fromDecimals` could be used here

**Review by tumppi — APPROVED** [review]
Left a couple of comments. Otherwise looking good 👍

File: src/hooks/useJoinSponsorship.ts (hunk 2 above)
Comment (pmichalowski) [reply src/hooks/useJoinSponsorship.ts:32]: The try/catch here is for the case of closing the modal, which is a promise rejection. As for the possible errors, they are being handled and displayed :)

File: src/hooks/useJoinSponsorship.ts (hunk 3 above)
Comment (pmichalowski) [reply src/hooks/useJoinSponsorship.ts:27]: Wel actually `freeFundsWei` shows invalud value :) it doesn't subtract the staked amount - that's why I did the math here. I'll raise the question about it on today's status call.

Comment (tumppi) [reply src/hooks/useJoinSponsorship.ts:27]: Hmm, weird. For me it seems to be working. I tried it with operator delegate/undelegate and it shows correctly poolvalue subtracted by amount staked in sponsorships.

**Pull request merged**